File: 0238-product-of-array-except-self/0238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const n = nums.length;
    const answer = new Array(n);
    const prefix = new Array(n).fill(1);
    const suffix = new Array(n).fill(1);
    
    // create a prefix array
    // prefix array will have the product of all element before an index i 
    for(let i = 1; i < n; i++) {
        prefix[i] = prefix[i - 1] * nums[i - 1];
    }
    
    // create a suffix array 
    // suffix array will have the prodict of all element before an index i 
    for(let i = n - 2; i >= 0; i--) {
        suffix[i] = suffix[i + 1] * nums[i + 1];
    }
    
    // populate the answers array 
    for(let i = 0; i < n; i++) {
        answer[i] = prefix[i] * suffix[i];
    }
    
    return answer;
};

// tc -> O(n)
// sc -> O(n)

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.go
func productExceptSelf(nums []int) []int {
    res := make([]int, len(nums))
    
    // calculate prefix
    prefix := 1
    for i, v := range nums {
        res[i] = prefix
        prefix *= v
    }
    
    // calculate postfix 
    postfix := 1
    for i := len(nums) - 1; i >= 0; i-- {
        res[i] *= postfix
        postfix *= nums[i]
    }
    
    return res
}